impeller/renderer/backend/vulkan/blit_command_vk.cc

```diff
@@ -0,0 +1,178 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/renderer/backend/vulkan/blit_command_vk.h"
+
+#include "impeller/renderer/backend/vulkan/commands_vk.h"
+#include "impeller/renderer/backend/vulkan/texture_vk.h"
+
+namespace impeller {
+
+BlitEncodeVK::~BlitEncodeVK() = default;
+
+// BEGIN: BlitCopyTextureToTextureCommandVK
+
+BlitCopyTextureToTextureCommandVK::~BlitCopyTextureToTextureCommandVK() =
+    default;
+
+std::string BlitCopyTextureToTextureCommandVK::GetLabel() const {
+  return label;
+}
+
+[[nodiscard]] bool BlitCopyTextureToTextureCommandVK::Encode(
+    FencedCommandBufferVK* fenced_command_buffer) const {
+  // cast source and destination to TextureVK
+  auto& source_tex_vk = TextureVK::Cast(*source);
+  auto& dest_tex_vk = TextureVK::Cast(*destination);
+
+  // get the vulkan image and image view
+  auto source_image = source_tex_vk.GetImage();
+  auto dest_image = dest_tex_vk.GetImage();
+
+  // copy the source image to the destination image, from source_region to
+  // destination_origin.
+  vk::ImageCopy image_copy{};
+  image_copy.setSrcSubresource(
+      vk::ImageSubresourceLayers(vk::ImageAspectFlagBits::eColor, 0, 0, 1));
+  image_copy.setDstSubresource(
+      vk::ImageSubresourceLayers(vk::ImageAspectFlagBits::eColor, 0, 0, 1));
+
+  image_copy.srcOffset =
+      vk::Offset3D(source_region.origin.x, source_region.origin.y, 0);
+  image_copy.dstOffset =
+      vk::Offset3D(destination_origin.x, destination_origin.y, 0);
+  image_copy.extent =
+      vk::Extent3D(source_region.size.width, source_region.size.height, 1);
+
+  // get single use command buffer
+  auto copy_cmd = fenced_command_buffer->GetSingleUseChild();
+
+  vk::CommandBufferBeginInfo begin_info;
+  begin_info.setFlags(vk::CommandBufferUsageFlagBits::eOneTimeSubmit);
+  auto res = copy_cmd.begin(begin_info);
+
+  if (res != vk::Result::eSuccess) {
+    VALIDATION_LOG << "Failed to begin command buffer: " << vk::to_string(res);
```

Comment:
The other cases where we return early due to failure also need a validation log detailing the failure.